modules/statistics/AnalyticsAdapter.js

```diff
@@ -1,130 +1,297 @@
+            };
+        } else if (typeof eventName === 'object') {
+            event = eventName;
+        }
+
+        if (!AnalyticsAdapter._verifyRequiredFields(event)) {
+            logger.error(
+                `Dropping a mis-formatted event: ${JSON.stringify(event)}`);
 
-        this.analyticsHandlers.add(cacheAnalytics);
+            return;
+        }
+
+        this._sendEvent(event);
     }
 
     /**
-     * Sends analytics event.
-     * @param {String} eventName the name of the event
-     * @param {Object} data can be any JSON object
+     * Checks whether an event has all of the required fields set, and tries
+     * to fill in some of the missing fields with reasonable default values.
+     * Returns true if after this operation the event has all of the required
+     * fields set, and false otherwise (if some of the required fields were not
+     * set and the attempt to fill them in with a default failed).
+     *
+     * @param event the event object.
+     * @return {boolean} true if the event (after the call to this function)
+     * contains all of the required fields, and false otherwise.
+     * @private
      */
-    sendEvent(eventName, data = {}) {
-        const modifiedData = Object.assign({}, this.permanentProperties, data);
-
-        this.analyticsHandlers.forEach(
-            analytics =>
-                analytics.sendEvent(
-                    eventName,
-                    analytics === cacheAnalytics ? data : modifiedData
-                )
-        );
+    static _verifyRequiredFields(event) {
+        if (!event) {
+            return false;
+        }
+
+        const type = event.type;
+
+        if (type !== TYPE_OPERATIONAL && type !== TYPE_PAGE
+            && type !== TYPE_UI && type !== TYPE_TRACK) {
+            logger.error(`Unknown event type: ${type}`);
+
+            return false;
+        }
+
+        if (type === TYPE_PAGE) {
+            return Boolean(event.name);
+        }
+
+        // Try to set some reasonable default values in case some of the
+        // parameters required by the handler API are missing.
+        event.action = event.action || event.name || event.actionSubject;
+        event.actionSubject = event.actionSubject || event.name || event.action;
+        event.source = event.source || event.name || event.action
+            || event.actionSubject;
+
+        if (!event.action || !event.actionSubject || !event.source) {
+            logger.error(
+                'Required field missing (action, actionSubject or source)');
+
+            return false;
+        }
+
+        // Track events have additional required fields.
+        if (type === TYPE_TRACK) {
+            event.objectType = event.objectType || 'generic-object-type';
+            event.containerType
+                = event.containerType || 'generic-container-type';
+
+            if (!event.objectType || !event.objectId
+                || !event.containerType || !event.containerId) {
```

Comment:
I think maybe the objectType and containerType checks aren't necessary because of the hardcoded fallbacks above.